File: LeetCode/permutations.cpp
// https://leetcode.com/problems/permutations/
// https://www.youtube.com/watch?v=GCm7m5671Ps&t=1446s
class Solution
{
public:
    void helper(vector<vector<int>> &res, int cur, vector<int> &nums)
    {
        if (cur == nums.size())
        {
            res.push_back(nums);
            return;
        }
        for (int i = cur; i < nums.size(); ++i)
        {
            swap(nums[i], nums[cur]);
            helper(res, cur + 1, nums);
            swap(nums[i], nums[cur]);
        }
        return;
    }
    vector<vector<int>> permute(vector<int> &nums)
    {
        vector<vector<int>> res;
        int n = nums.size();
        helper(res, 0, nums);
        return res;
    }
};

File: LeetCode/permutations.java
// https://leetcode.com/problems/permutations/

class Solution {
  public void backtrack(List<List<Integer>> result, List<Integer> tmp, int[] nums, Set<Integer> done) {
    if (tmp.size() == nums.length) {
      result.add(new ArrayList<>(tmp));
      return;
    }

    for (int i = 0; i < nums.length; ++i) {
      if (done.contains(nums[i])) {
        continue;
      }
      done.add(nums[i]);
      tmp.add(nums[i]);

      backtrack(result, tmp, nums, done);

      done.remove(nums[i]);
      tmp.remove(tmp.size() - 1);
    }
  }

  public List<List<Integer>> permute(int[] nums) {
    List<List<Integer>> result = new ArrayList();
    backtrack(result, new ArrayList<Integer>(), nums, new HashSet<>());
    return result;
  }
}